keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/JVMDoubleTensor.java

```diff
@@ -0,0 +1,1740 @@
+        boolean resultShapeIsLeftSideShape = Arrays.equals(resultShape, paddedLeftShape);
+
+        final double[] outputBuffer;
+        if (!resultShapeIsLeftSideShape) {
+
+            boolean resultShapeIsRightSideShape = Arrays.equals(resultShape, paddedRightShape);
+
+            if (!resultShapeIsRightSideShape) {
+                throw new IllegalArgumentException(
+                    "Broadcasting of shape " + Arrays.toString(paddedLeftShape) + " and " + Arrays.toString(paddedRightShape) + " not supported."
+                );
+            }
+
+            outputBuffer = new double[Ints.checkedCast(TensorShape.getLength(resultShape))];
+
+        } else {
+            outputBuffer = inPlace ? leftBuffer : new double[leftBuffer.length];
+        }
+
+        //Allow broadcasting from left and right
+        if (paddedLeftShape.length > paddedRightShape.length || leftBuffer.length > rightBuffer.length) {
+            //e.g. [2, 2] * [1, 2]
+            broadcastFromRight(leftBuffer, paddedLeftShape, paddedLeftStride, rightBuffer, paddedRightShape, paddedRightStride, outputBuffer, op);
+        } else {
+            //e.g. [2] / [2, 2]
+            broadcastFromLeft(leftBuffer, paddedLeftShape, paddedLeftStride, rightBuffer, paddedRightShape, paddedRightStride, outputBuffer, op);
+        }
+
+        if (inPlace) {
+            left.buffer = outputBuffer;
+            left.shape = resultShape;
+
+            //TODO: get rid of the need to do this!
+            left.stride = TensorShape.getRowFirstStride(resultShape);
+
+            return left;
+        } else {
+            return new JVMDoubleTensor(outputBuffer, resultShape);
+        }
+    }
+
+    private JVMDoubleTensor elementwiseBinaryOp(JVMDoubleTensor left, double[] leftBuffer, double[] rightBuffer, long[] shape,
+                                                BiFunction<Double, Double, Double> op,
+                                                boolean inPlace) {
+
+        final double[] outputBuffer = inPlace ? leftBuffer : new double[leftBuffer.length];
+
+        for (int i = 0; i < outputBuffer.length; i++) {
+            outputBuffer[i] = op.apply(leftBuffer[i], rightBuffer[i]);
+        }
+
+        if (inPlace) {
+            left.buffer = outputBuffer;
+            return left;
+        } else {
+            return new JVMDoubleTensor(outputBuffer, shape);
+        }
+    }
+
+    @Override
+    public FlattenedView<Double> getFlattenedView() {
+        return new JVMDoubleFlattenedView();
+    }
+
+    private class JVMDoubleFlattenedView implements FlattenedView<Double> {
+
+        @Override
+        public long size() {
+            return buffer.length;
+        }
+
+        @Override
+        public Double get(long index) {
+            return buffer[Ints.checkedCast(index)];
+        }
+
+        @Override
+        public Double getOrScalar(long index) {
+            if (buffer.length == 1) {
```

Comment:
possible optimisation: move this check into `getFlattenedView`, and based on its result return one of 2 different classes `ScalarJVMDoubleFlattenedView` or `TensorJVMDoubleFlattenedView`